src/Symfony/Component/Messenger/Asynchronous/Routing/AbstractSenderLocator.php

```diff
@@ -11,28 +11,34 @@
 
 namespace Symfony\Component\Messenger\Asynchronous\Routing;
 
+use Symfony\Component\Messenger\Envelope;
+
 /**
  * @author Samuel Roze <samuel.roze@gmail.com>
  *
  * @internal
  */
 abstract class AbstractSenderLocator implements SenderLocatorInterface
 {
-    public static function getValueFromMessageRouting(array $mapping, $message)
+    public static function getValueFromMessageRouting(array $mapping, Envelope $envelope)
     {
-        if (isset($mapping[\get_class($message)])) {
-            return $mapping[\get_class($message)];
-        }
-        if ($parentsMapping = array_intersect_key($mapping, class_parents($message))) {
-            return current($parentsMapping);
+
+        if (isset($mapping[$class = \get_class($envelope->getMessage())])) {
+            return $mapping[$class];
         }
-        if ($interfaceMapping = array_intersect_key($mapping, class_implements($message))) {
-            return current($interfaceMapping);
+
+        foreach (class_parents($class) as $name) {
+            if (isset($mapping[$name])) {
+                return $mapping[$name];
+            }
         }
-        if (isset($mapping['*'])) {
-            return $mapping['*'];
+
+        foreach (class_implements($class) as $name) {
```

Comment:
`false` as 2nd argument?

Reply:
the class is already loaded, adding this argument doesn't save anything